### DISCUSSION TITLE
Simulate big models problem

### DISCUSSION THREAD

**Discussion by andreili** [Xyce/Xyce#90] [Q&A]
Hi, all!
I try to simulate a simple SoC model on SPICE, but get a veeery slow calculation.
Firstly, I try to simulate a single block. Problem size - 47863 (4.79E+4). Best solver - KLU2:
| Solver | Time |
|--------|--------|
| default | 1177 |
| KLU | 500 |
| KLU2 | 409 |
| KSparse | 426 |
| AztecOO | 1548 | 
| Belos | 1848 |

On this simulation, I see a 88% CPU loading, but in fact - 100% (any games laged and freezed).
And next simulation - full model. problem size - 528800 (5.29E+5). Try to run with a KLU2 solver and see a 16..28% CPU load and loadavg above 2.5. But, in fact - CPU a 100% loaded! Any application shown a freezes (more noticeable in any game).

Why such a gap in displayed and actual CPU usage? If necessary, I can upload files to run this simulation (only the simulator itself is required).

**Comment by Karlsefni2012** [comment]
Are you running in serial or parallel?

**Reply by andreili** [reply]
> Are you running in serial or parallel?

@Karlsefni2012 Parallel - see a set of threads, each thread eat part of CPU's time :)
On smallest problem - each thread utilize more CPU time...
![image](https://github.com/Xyce/Xyce/assets/1331777/21dfc7a7-9bd0-4c84-ae84-36adb53f9e25)

**Reply by andreili** [reply]
@Karlsefni2012 Yes, you alright - it's a serial run. But with mutithreading.
I try t run a parallel run (mpirun -np 4 Xyce) - shows a same result for CPU load (see a image for small problem, 4E+4).
And I try to start a parallel run for bigest problem...
![image](https://github.com/Xyce/Xyce/assets/1331777/03545f63-c977-4d5f-b43f-9a863603f8c9)

**Comment by hkthorn** [comment]
@andreili So, at this time, Xyce does not have any shared memory parallelism in it and cannot take advantage of multiple threads in a serial build.  The only exception would be if the linear solver was multithreaded, or the underlying math libraries were multithreaded, but the list of solvers above are not multithreaded.   I would suggest turning off multithreading for the serial run.

The performance of Xyce is largely dictated by two parts of the code: (1) the evaluation of the devices in the circuit and (2) the linear solver used within Newton's method.  For circuits where the linear system is order 40k, distributed memory parallelism (MPI) might provide some speedup in the evaluation of the devices, but not in the linear system solve.   Around 100k, you might start to see improvement in the linear system solve through iterative methods (AztecOO/Belos) if there is an effective preconditioner, which is problem dependent.  Again, shared memory parallelism through multithreading will only impact the linear solve, if the solver is multithreaded.  There's no indication that a thread-aware linear solver is being used here, so I would suggest turning off multithreading in parallel as well.

**Reply by andreili** [reply]
@hkthorn Multithreaded solver - from Trilinos library, with default options. And it's a really speed-up. I compare with single-threaded, and it's a more than x15 slowly between multithreaded variant.
**MPI parallel run** have a same problem - low CPU loading. Now, I wait finishing a initialization parallel run (e.g. mpirun -n 4) for a big simulation - and wait a 30 minutes for first results.

**Reply by andreili** [reply]
Hmmm, interesting...
See a picture - it's a CPU loading graph :)
First part - "reading and parsing netlist" stage, all Ok.
Second part (with a stable loading) - "Initializating" stage.
And third stage - "DC Operating Point Calculation", with a with spikes in consumption (CPU and memory). Memory consumption of a main task fluctuates between 700MB and 13GB every few second (synchronous to CPU consumption).
And this solution is a veeery slow - after 10 minutes of running, I can't see a "1% completed" message, nothing.
On serial run for this simulation - need only a 5 minutes to simulation (from terminal command to finish), wuth a 70..800MB memory and ~67% CPU consumptions.
For a parallel run - command "mpirun -n 10", for experiment.
![image](https://github.com/Xyce/Xyce/assets/1331777/36c6cf5d-f72b-4040-a38c-8999b37ef989)
P.S.: I check a RAW-file (to save a simulation result for every step) - and it's a very small and contain a first few step only for 16 minutes of running... It's a fail...

**Reply by andreili** [reply]
For a parallel run (mpirun -n 4) see a similar result - very slow solution. For a 20 minutes run, proceed only a 28 points.
I terminate it and run a serial run, without changes on models/params - it's took a 263 second fro fully simulation (954 points)...

**Reply by Karlsefni2012** [reply]
Exactly which multithreaded solver are you using?  None of the ones you listed are multithreaded.

Some of them (the iterative solvers like Belos and AztecOO) are MPI-parallel.

As I mentioned in my other comments, there are threaded solvers in Trilinos.

**Reply by Karlsefni2012** [reply]
During setup the initial setup. even when running in parallel there are phases of the initial setup that are serial.   The code can only be "fully parallel" once the setup is over and the actual calculation begins.

But it is hard to really make much comment without more detail.  Obviously Xyce is struggling to find a solution, for one reason or another.  Is this a circuit you can share?

Are you running a build of Xyce that you built yourself? or did you download one of our binaries?   

The problem size you reported, 47863 is small enough that one of the direct solvers (like KLU or KLU2) is almost certainly the best choice.  That is small enough that the iterative solvers are unlikely to do better.   

If you have a serial binary, it will by default use KLU for its linear solve, regardless of problem size.

If you have a parallel binary, it will use KLU if the problem size is very small (< 10,000 I think), and otherwise will by default use AztecOO.   But even for 47863, while bigger than 10k, is small by iterative solver standards.  So, to run well in parallel you should still use KLU (or KLU2 if that works better).

But of course, this will mean that 1/2 of the calculation (linear solve) is in serial and the other 1/2  (device evaluation) is in parallel.  So, even if the problem solves successfully, the amount of parallel speedup you will get will be limited by Ahmdal's law.     Also, for any parallel configuration (parallel solve or not) there will always be diminishing returns for adding more processors for a fixed problem size.   

Assuming you are running in parallel using KLU, then a 4 process run will be faster than a 1 processor run, but by < 4x.    You can get an estimate from the Amdal's law formula.  (https://en.wikipedia.org/wiki/Amdahl%27s_law)

But of course, if the solvers are struggling to converge (which seems to be the case with your circuit) running in parallel won't fix that problem.

**Reply by andreili** [reply]
@Karlsefni2012 Yes, I can a share a circuit - It's a my Open-Source project :)
https://drive.google.com/file/d/1sUZE8_0iELUDbS7-GYgm9I-1VWLVcTGw/view?usp=sharing
Small problem - on dir "blocks/pc_inc", file "pc_inc_sim.spice". "Problem size" is 3272.
Medium problem - on dir "blocks/rv_fetch_buf", file "rv_fetch_buf_sim.spice". "Problem size" is 47863.
Big problem - on dir "blocks/top", file "top_sim.spice". "Problem size" is 528800.


All solvers from first message running in threaded mode on serial run of Xyce (thread count - equal to maximum CPU thread count). I use Trilinos with custom compilation parameters (default from gentoo and minimal requrinments for Xyce) and default build of Xyce (only "Xyce_PARALLEL_MPI=True" is specified).
And for experiment - I compare with Trilinos v13.4.1 and v14.0.0 (need  some tricks on Xyce build files). All working, with best speed on serial runs for v14.0.0.
Build options for Trilinos:
`cmake -G "Unix Makefiles" -DCMAKE_INSTALL_PREFIX=/usr -DBUILD_SHARED_LIBS=ON -DCMAKE_INSTALL_PREFIX= -DCMAKE_SKIP_INSTALL_RPATH=ON -DCMAKE_INSTALL_RPATH_USE_LINK_PATH=OFF -DTrilinos_INSTALL_CONFIG_DIR=/usr/lib64/cmake -DTrilinos_INSTALL_INCLUDE_DIR=/usr/include/trilinos -DTrilinos_INSTALL_LIB_DIR=/usr/lib64/trilinos -DTrilinos_ENABLE_ALL_PACKAGES=no -DTrilinos_ENABLE_Adelus=OFF -DTrilinos_ENABLE_Moertel=OFF -DTrilinos_ENABLE_PyTrilinos=OFF -DTrilinos_ENABLE_SEACAS=OFF -DTrilinos_ENABLE_Amesos=ON -DTrilinos_ENABLE_AztecOO=ON -DTrilinos_ENABLE_EpetraExt=ON -DTrilinos_ENABLE_Epetra=ON -DTrilinos_ENABLE_Ifpack=ON -DTrilinos_ENABLE_ML=ON -DTrilinos_ENABLE_MueLu=ON -DTrilinos_ENABLE_OpenMP=yes -DTrilinos_ENABLE_ROL=ON -DTrilinos_ENABLE_Sacado=ON -DTrilinos_ENABLE_Teuchos=ON -DTrilinos_ENABLE_Tpetra=ON -DTrilinos_ENABLE_Zoltan=ON -DTrilinos_ENABLE_TESTS=no -DTPL_ENABLE_BinUtils=ON -DTPL_ENABLE_BLAS=ON -DTPL_ENABLE_LAPACK=ON -DTPL_ENABLE_MPI=ON -DTPL_ENABLE_ADOLC=no -DTPL_ENABLE_AMD=yes -DTPL_ENABLE_ARPREC=no -DTPL_ENABLE_BLACS=no -DTPL_ENABLE_BoostLib=ON -DTPL_ENABLE_Boost=ON -DTPL_ENABLE_Clp=no -DTPL_ENABLE_CSparse=yes -DTPL_ENABLE_CUDA=no -DTPL_ENABLE_CUSPARSE=no -DTPL_ENABLE_Eigen=no -DTPL_ENABLE_GLPK=no -DTPL_ENABLE_gtest=no -DTPL_ENABLE_HDF5=no -DTPL_ENABLE_HWLOC=no -DTPL_ENABLE_HYPRE=no -DTPL_ENABLE_Matio=no -DTPL_ENABLE_METIS=no -DTPL_ENABLE_MKL=no -DTPL_ENABLE_MUMPS=no -DTPL_ENABLE_Netcdf=no -DTPL_ENABLE_PARDISO_MKL=no -DTPL_ENABLE_PETSC=no -DTPL_ENABLE_Pthread=yes -DTPL_ENABLE_QD=no -DTPL_ENABLE_SCALAPACK=no -DTPL_ENABLE_Scotch=no -DTPL_ENABLE_SuperLU=yes -DTPL_ENABLE_TAUCS=yes -DTPL_ENABLE_TBB=no -DTPL_ENABLE_Thrust=no -DTPL_ENABLE_TVMET=no -DTPL_ENABLE_UMFPACK=yes -DTPL_ENABLE_X11=yes -DTPL_ENABLE_yaml-cpp=no -DTPL_ENABLE_Zlib=yes -DTrilinos_ENABLE_TrilinosCouplings=ON -DEpetraExt_BUILD_BTF=ON -DEpetraExt_BUILD_EXPERIMENTAL=ON -DEpetraExt_BUILD_GRAPH_REORDERINGS=ON -DTrilinos_ENABLE_COMPLEX=ON -DAmesos_ENABLE_CSparse=yes -DTrilinos_ENABLE_ShyLU_NodeTacho=OFF -DAmesos_ENABLE_ParMETIS=ON -DAmesos2_ENABLE_Basker=ON -DCMAKE_CXX_FLAGS:STRING=-DSHYLU_NODEBASKER -DAmesos2_ENABLE_LAPACK=ON -DAmesos2_ENABLE_MUMPS=OFF -DTpetra_INST_SERIAL=ON -DCSparse_LIBRARY_NAMES=cxsparse -DBLAS_LIBRARY_NAMES=blas -DLAPACK_LIBRARY_NAMES=lapack -DSuperLU_LIBRARY_NAMES=superlu;blas -DSuperLU_INCLUDE_DIRS=/usr/include/superlu -DCMAKE_BUILD_TYPE=Release`

**Reply by Karlsefni2012** [reply]
Thanks for posting the circuits.  We can take a look at those.   I am not sure how quickly we can get to it, but we'll do it.

Using more recent versions of Trilinos is mostly fine.  The only reason that our "official" build process is still stuck on Trilinos 12.12 has to do with some of the details of our automated testing.  Once that gets fixed the "official" build will be with a version that is more recent.  Internally, several of us routinely build with more recent versions.

Regarding threading, just turing on "openMP" in trilinos isn't likely to buy you much with Xyce.   Or, at least, it shouldn't.  However, if you are seeing some speedups that is interesting.

As @hkthorn mentnioned, there are two major aspects of Xyce's (or any SPICE simulator's) transient calculation.  One is the device evaluation, and the other is the linear solve.  Xyce's device evaluation does not use Trilinos directly and it isn't instrumented for threading.  So, threaded-enabled  Trilinos shouldn't have much (any?) impact on that phase.   

For the linear solve, only some of the linear solvers are designed to use threading.  If you aren't using one of those, then having threading  enabled won't benefit that phase either.  KLU, for example, is an old solver from 20 years ago and doesn't have threading built into it.  

So, if (for example) you're using threaded Trilinios and using large number of threads, I can't say what is happening on all those threads.  Possibly there are some low-level vector operations that are taking advantage of threading, but I wouldn't expect that to be a significant part of the calculation.     I've seen that sort of thing when I've experimented with threaded BLAS libraries in the past.

**Reply by tvrusso** [reply]
Just as a minor aside on top of all the other things that have been said so far.  This is separate from enabling threading in Trilinos (which has already been discussed sufficiently), but relates to several comments above regarding the potential for some gain by using threaded BLAS and LAPACK libraries.

There are some threaded BLAS and LAPACK libraries that exhibit runaway thread usage with Xyce sometimes, which is one of the reasons we usually don't suggest using them and our build guidance hasn't traditionally included instructions for them.  

ATLAS, for example, will occasionally try to spawn as many threads as the machine has cores, and with Xyce we have often seen it be less efficient than straight serial (sometimes dramatically so) when it does this.  I used to build Xyce with ATLAS myself a lot, and gave it up because of this behavior.

The same thing occurs with the threaded Intel MKL library.  We have never tracked down what happens in those rare cases where it seems to run away and spin its wheels doing who-knows-what on as many threads as the machine can support, but it does sometimes.  We have seen it often, and in the Xyce testing process we typically set environment variables to tell the MKL to limit the number of threads it's allowed to spawn to one or two instead of "as many as you want."

This is compounded if one is running more than one Xyce process, or is running Xyce in MPI parallel (i.e., more than one Xyce process, but working together).  The libraries are not smart enough to restrict their thread numbers in a way that takes into account that there are other processes doing the same thing --- so you can get multiple Xyce processes *each* trying to have one thread for every core on the machine, oversubscribing the machine and leading to worse performance than if one hadn't tried threading at all.

This comment is little more than an additional statement repeating what @Karlsefni2012 and @hkthorn have already said: that other than by using specific (and somewhat experimental) non-default multithreaded solvers included in Trilinos, Xyce is not set up to take good advantage of multithreading, and if you want to distribute your problem across multiple processors efficiently, the message passing implementation in a parallel build of Xyce is the way to do it.

**Reply by andreili** [reply]
@Karlsefni2012 @tvrusso 
I tried running the biggest problem with different solvers and in different modes (serial and parallel). The results roughly coincided with what can be seen in the very first message. And all next runs - only a serial mode. Parallel run in my case is a extremely slowly.
Belos/Basker - strictly single-threaded.
KLU, KLU2, KSparse, AztecOO, Belos - in my version of the library assembly, all are multi-threaded. And only in AztecOO and Belos multithreading makes no sense - the speed has not increased at all.
And the results of the rest of the solvers were significantly accelerated - when using 20 threads (i5-13500), the average speedup reached about 4 times.

**Comment by Karlsefni2012** [comment]
Xyce isn't designed to make use of threading.   There are some exceptions to this, but mostly you should consider Xyce's parallelism to be based on message-passing.

The exceptions include (1) building Xyce using a threaded BLAS library and (2) invoking one of the linear solvers that actually makes use of threading.   I've only done (1) as an experiment and I'm not sure it helps very much.  Regarding (2), none of the solvers you've tried are instrumented for threading.    The main solvers inside of trilinos that are instrumented for threading include Basker, SuperLU and Pardiso.

**Reply by Karlsefni2012** [reply]
Just realized that I was writing a redundant comment, with @hkthorn
